File: Day #809/solution.cpp
// Day 809: Check balanced round/curly/square brackets using a stack.
// Push opens, match closes against stack top. Time O(N), Space O(N).
#include <bits/stdc++.h>
using namespace std;

bool isBalanced(const string& s) {
    stack<char> st;
    unordered_map<char, char> match = {{')', '('}, {']', '['}, {'}', '{'}};
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') st.push(c);
        else if (match.count(c)) {
            if (st.empty() || st.top() != match[c]) return false;
            st.pop();
        }
    }
    return st.empty();
}

int main() {
    cout << boolalpha;
    cout << isBalanced("([])[]({})") << "\n"; // true
    cout << isBalanced("([)]") << "\n";        // false
    cout << isBalanced("((()") << "\n";        // false
    return 0;
}

File: Day #809/solution.js
// Day 809: Check balanced round/curly/square brackets using a stack.
// Push opens, match closes against stack top. Time O(N), Space O(N).

function isBalanced(s) {
  const stack = [];
  const match = { ")": "(", "]": "[", "}": "{" };
  for (const c of s) {
    if (c === "(" || c === "[" || c === "{") stack.push(c);
    else if (c in match) {
      if (stack.pop() !== match[c]) return false;
    }
  }
  return stack.length === 0;
}

console.log(isBalanced("([])[]({})")); // true
console.log(isBalanced("([)]"));        // false
console.log(isBalanced("((()"));        // false
